src/EnergyPlus/HVACFourPipeBeam.cc

```diff
@@ -0,0 +1,1302 @@
+// C++ Headers
+#include <cmath>
+
+// ObjexxFCL Headers
+#include <ObjexxFCL/Array.functions.hh>
+#include <ObjexxFCL/Fmath.hh>
+
+// EnergyPlus Headers
+#include <AirTerminalUnit.hh>
+#include <HVACFourPipeBeam.hh>
+#include <BranchNodeConnections.hh>
+#include <DataContaminantBalance.hh>
+#include <DataDefineEquip.hh>
+#include <DataEnvironment.hh>
+#include <DataHVACGlobals.hh>
+#include <DataIPShortCuts.hh>
+#include <DataLoopNode.hh>
+#include <DataPlant.hh>
+#include <DataPrecisionGlobals.hh>
+#include <DataSizing.hh>
+#include <DataZoneEnergyDemands.hh>
+#include <DataZoneEquipment.hh>
+#include <FluidProperties.hh>
+#include <General.hh>
+#include <GeneralRoutines.hh>
+#include <InputProcessor.hh>
+#include <NodeInputManager.hh>
+#include <OutputProcessor.hh>
+#include <PlantUtilities.hh>
+#include <Psychrometrics.hh>
+#include <ReportSizingManager.hh>
+#include <ScheduleManager.hh>
+#include <UtilityRoutines.hh>
+#include <CurveManager.hh>
+
+
+namespace EnergyPlus {
+
+namespace FourPipeBeam {
+
+	Array1D< std::shared_ptr< HVACFourPipeBeam > > FourPipeBeams; // dimension to number of machines
```

Comment:
@Myoldmopar An entity that doesn't have a natural owner can indeed be a natural use of `shared_ptr`. While we are still having "master" containers we don't need shared ownership: an owning (unique or raw) pointer in the master container and exposing only via reference in the public interface works fine. As we move away from master containers we may need some shared ownership but using them everywhere is a common mistake: they create problems beyond performance and, unfortunately, they aren't a magic wand to make memory management easy. I encourage reading up on the drawbacks of shared pointers followed by someone writing up a proposed guideline on memory/ownership and a discussion. Until then I suggest that any use of `shared_ptr` must be justified and alternatives evaluated. The problems with "works great so far" are that we aren't assessing performance, code will be used as a model by the next developer, and shared ownership pointer issues snowball as they spread out in large projects: this really needs to be thought through with care before we go too far. Using shared pointers widely and waiting until we find we've created a problem is a mistake.